src/optimization/hooke_jeeves_method.cpp

```diff
@@ -0,0 +1,161 @@
+{
+    vector<double> arguments(initial_parameters);
+    vector<double> arguments_after_search(initial_parameters.size());
+    vector<double> steps_array = vector<double>(arguments.size(), init_step_);
+    size_t iteration = 0;
+
+    do
+    {
+        arguments_after_search = CoordinatewiseSearch(arguments, steps_array);
+        if (common::FloatingPoint::AreEqual(arguments, arguments_after_search))
+        {
+            for (size_t i = 0; i < steps_array.size(); ++i)
+            {
+                steps_array[i] /= step_divider_;
+            }
+            continue;
+        }
+
+        vector<double> next_step_arguments(arguments.size());
+        vector<double> next_step_arguments_after_search(arguments.size());
+        while (true)
+        {
+            for (size_t i = 0; i<arguments.size(); ++i)
+            {
+                next_step_arguments[i] = arguments[i] + acceleration_
+                        * (arguments_after_search[i] - arguments[i]);
+            }
+            next_step_arguments_after_search = CoordinatewiseSearch(next_step_arguments, steps_array);
+
+            arguments = arguments_after_search;
+            if (CountObjectiveFunctionValueToMinimize(arguments_after_search) <=
+                    CountObjectiveFunctionValueToMinimize(next_step_arguments_after_search))
+            {
+                break;
+            }
+            arguments_after_search = next_step_arguments_after_search;
+
+            iteration++;
+            // Check number of iterations
+            if (iteration == max_iteration_number_)
+            {
+                std::cout << "Solution not found" << std::endl;
+                return arguments_after_search;
+            }
+        }		
+    } while (!IsExitCriteriaFulfilled(steps_array));
+    return arguments_after_search;
+}
+
+bool HookeJeevesMethod::IsExitCriteriaFulfilled(
+    const vector<double>& steps_array) const
+{
+    double temp = 0;
+    // check if step is small enough
+    for (size_t j = 0; j<steps_array.size(); ++j)
+    {
+        temp += steps_array[j] * steps_array[j];
+    }
+    if (std::sqrt(temp) <= precision_)
+    {
+        return true;
+    }
+    return false;
+}
+
+vector<double> HookeJeevesMethod::CoordinatewiseSearch(
+    const vector<double>& arguments,
+    const vector<double>& steps_array) const
+{
+    vector<double> new_arguments(arguments);
+    vector<double> temp(arguments);
+    for (size_t i = 0; i < arguments.size(); ++i)
+    {
+        double new_function_value = CountObjectiveFunctionValueToMinimize(new_arguments);
+        temp[i] = new_arguments[i] + steps_array[i];
+        if (CountObjectiveFunctionValueToMinimize(temp) >= new_function_value)
+        {
+            temp[i] = new_arguments[i] - steps_array[i];
+            if (CountObjectiveFunctionValueToMinimize(temp) >= new_function_value)
+            {
+                temp[i] = new_arguments[i];
```

Comment:
Not very clear logic with temp and new_arguments values, which are sometimes reset to their initial values and sometimes new values are applied. This should be just 2 tests for positive and negative step, this should be more clear. 
Try to refactor to achieve more readable code - it should be self-explanatory. Do not afraid to copy whole vectors where necessary (not reset sole indexes) if necessary - just assign one vector variable to another then.